### ISSUE TITLE
CA1508 False Positive on variable modified in closure

### ISSUE THREAD

**Issue by j2jensen** [dotnet/roslyn-analyzers#4145]
### Analyzer

**Diagnostic ID**: [CA1508](https://docs.microsoft.com/visualstudio/code-quality/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.3.0 (Latest)

### Describe the bug

When only closures set variable values, the code analyzer assumes that the variable never gets assigned.

### Steps To Reproduce
Compile the following methods.

        private static bool Test()
        {
            object o = null;
            Do(() => o = new object());
            return o == null;
        }

        private static void Do(Action a) => a();

### Expected behavior
The code analyzer should ignore this case because it doesn't know for certain that the variable isn't assigned.

### Actual behavior
The code analyzer matches on `o == null`

**Comment by SebastianSchumann** [comment]
I'm not sure that this should _not_ show a warning. Due to deferred exectution it is hard to decide whether or not  the instance is assigned.

If a warning appears every developer can decide to suppress it if he's sure about a correct assignment.

IMHO: The most common case might be that this is a mistake and the warning is correct.

**Comment by mavasani** [comment]
@j2jensen The analyzer is reporting `'o == null' is always 'false'`, i.e. it is _always assigned_ to a non-null value on non-exception code paths.

**Comment by mavasani** [comment]
Also note, following code generates no warning:

```csharp
    private static bool Test(object o2)
    {
        object o = null;
        Do(() => o = o2);
        return o == null;
        void Do(Action a) => a();
    }
```

**Comment by j2jensen** [comment]
I apologize: I was in a bit of a hurry when trying to get a minimal repro. I've updated the repro steps above to make the `Do` method non-private, in order to more closely reproduce the exact issue I was seeing. Now it reports:

> 'o == null' is always 'true'.

![image](https://user-images.githubusercontent.com/453495/92663418-79d60a80-f2be-11ea-829f-b4e4d6509442.png)

@SebastianSchumann I can certainly understand the argument that maybe there should be some kind of warning in cases like these: something more along the lines of "you're modifying a variable in a closure". But I don't think this particular warning is accurate, since `o == null` is _not_ always `true`, and it would be dangerous to encourage developers to replace this expression with `true`.  

There are already warnings to tell you if a local variable is used before it's initialized:
![image](https://user-images.githubusercontent.com/453495/92663725-48aa0a00-f2bf-11ea-8bda-25676ef51adb.png)

**Comment by mavasani** [comment]
Thanks. Seems like the fact that interprocedural analysis is disabled by default leads to this false positive. It works correctly if you specify the below in your .editorconfig

```
[*.cs]
dotnet_code_quality.interprocedural_analysis_kind = ContextSensitive
```

Seems like we need to pessimistically clear state for all captured and assigned variables for calls that are not interprocedurally analyzed.

**Comment by j2jensen** [comment]
FYI, I just encountered a similar issue when the delegate I'm calling is async, and it doesn't appear to go away when setting `dotnet_code_quality.interprocedural_analysis_kind = ContextSensitive`.

**Comment by j2jensen** [comment]
I just verified that this error still occurs with the code I posted [here](https://github.com/dotnet/roslyn-analyzers/issues/4145#issuecomment-689874570), even with `dotnet_code_quality.interprocedural_analysis_kind = ContextSensitive` in `.editorconfig`.

I still feel like this is clearly a bug. If a variable is changed in a closure, even if the code can't guarantee the closure is ever executed, it's wrong to act like the variable will always have a given value. Following the analyzer's suggestion in something like an `if` condition could lead to people removing a bunch of code without realizing the variable actually was getting changed.
